File: 2023-24/Finales/2016-Febrero/Solución/eulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;
import dataStructures.set.Set;

import java.util.Iterator;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {

        Set<V> vertices = g.vertices();
        boolean es = true;

        for(V v : vertices){

            es = g.degree(v) % 2 == 0;
        }

        return es;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {

        g.deleteEdge(v,u);

        for(V vert : g.vertices()){
            if(g.degree(vert) == 0) g.deleteVertex(vert);
        }

    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {

        List<V> ciclo = new ArrayList<V>();
        ciclo.append(v0);

        Set<V> suc = g.successors(v0);
        Iterator<V> it = suc.iterator();
        V u = it.next();
        ciclo.append(u);
        remove(g, v0, u);

        while (u != ciclo.get(0) && !g.isEmpty() && !g.successors(u).isEmpty()) {
            suc = g.successors(u);
            it = suc.iterator();
            v0 = u;
            u = it.next();
            ciclo.append(u);
            remove(g, v0, u);
        }

        return ciclo;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
    		// TO DO
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
    		// TO DO
    		return null;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
    		// TO DO
        return null;
    }
}

File: 2023-24/Finales/2016-Febrero/Solución/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = aux (vertices g)
    where
        aux [] = True
        aux (x:xs) = even (degree g x) && aux xs

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v, u) = aux (deleteEdge g (v, u)) (vertices g)
  where
    aux grafo [] = grafo
    aux grafo (x:xs)
      | degree grafo x == 0 = aux (deleteVertex grafo x) xs
      | otherwise           = aux grafo xs

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux g v0 [v0]
    where
        aux g v xs
            | u == v0 = (g', u:xs)
            | otherwise = aux g' u (u:xs)
                where
                    g' = remove g (v,u)
                    u = head (successors g v)

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys     = ys
connectCycles (x:xs) (y:ys)
        | x == y    = (y:ys) ++ xs
        | otherwise = x : connectCycles xs (y:ys)

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g xs = head $ filter (\v -> v `elem` xs) (vertices g)

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g
    | not (isEulerian g) = error "Error: Grafo No Euleriano"
    | otherwise = aux g (head (vertices g)) []
        where
            aux g v xs
                | isEmpty g = xs
                | otherwise = aux g' (last xs') xs'
                    where
                        (g', cicloParcial) = extractCycle g v
                        xs' = connectCycles xs cicloParcial
